OpenSim/Common/DataTable.h

```diff
@@ -0,0 +1,274 @@
+//=============================================================================
+// METHODS
+//=============================================================================
+public:
+    virtual ~DataTable_(){};
+
+    /** Construct an empty DataTable */
+    DataTable_() : _type(typeid(DataType)) {}
+
+    /** Construct a pre-sized DataTable to be filled in later. If values
+        are accessed before they are assigned they will be NaN. */
+    DataTable_(size_t nrows, size_t ncols, 
+        const DataType& val = DataType(SimTK::NaN)) :
+        _data(int(nrows), int(ncols), val),
+        _type(typeid(DataType)) {}
+
+    /** Convenience construction of a DataTable from a filename with its source
+        type identified by the file extension. E.g. `results.csv`, `states.sto`
+    @param[in] filename         the name of the data file to be accessed
+    */
+    DataTable_(const std::string& filename) :
+        _type(typeid(DataType)) {
+        std::string daId = FileAdapter::findExtension(filename);
+        auto* reader = FileAdapter::createAdapter(daId);
+        if (!reader){
+            throw Exception("DataTable() no adapter to read file '"
+                + filename + "' could be found.");
+        }
+        reader->setFilename(filename);
+        reader->prepareForReading(*this);
+        reader->read();
+        delete reader;
+    }
+
+    DataTable_& operator=(const DataTable_& dt) {
+        _metaData = dt.getMetaData();
+        _data = dt.getAsMatrix() ;
+        _columnLabels = dt.getColumnLabels();
+        return *this;
+    }
+
+    virtual AbstractDataTable* clone() const override {
+        return new DataTable_(*this);
+    }
+
+    const XMLDocument& getMetaData() const override final { return _metaData; }
+    XMLDocument& updMetaData() override final { return _metaData; }
+
+    size_t  getNumRows() const override { return _data.nrow(); }
+    size_t  getNumCols() const override { return _data.ncol(); }
+
+    const Array<std::string>& getColumnLabels() const override {
+        return _columnLabels;
+    }
+
+    Array<std::string>& updColumnLabels() override {
+        return _columnLabels;
+    }
+    
+    /** Return true if the DataTable contains a column for the label. */
+    bool hasColumn(const std::string& label) const override{
+        return (_columnLabels.findIndex(label) >= 0);
+    }
+
+    size_t getColumnIndex(const std::string& label) const override {
+        return _columnLabels.findIndex(label);
+    }
+
+    /** Read-only access to a row of data in the table as a RowVector.*/
+    const SimTK::RowVectorView_<DataType> getRow(size_t row) const {
+        return _data[int(row)];
+    }
+
+    /** Write access to a column of data as a RowVector */
+    SimTK::RowVectorView_<DataType> updRow(size_t row) {
+        return _data[int(row)];
+    }
+
+    /** Read-only access to a column of data in the table as a Vector.*/
+    const SimTK::VectorView_<DataType> getColumn(size_t col) const {
```

Comment:
VectorView is identical to Vector except that it has shallow copy semantics meaning that when you assign something to it you get a view into the original data rather than a new copy. The "View" part does not imply read-only; that's managed by an internal flag.

The difference in naming between `Vector` and `RowVector` (and related classes) is because column vectors are so much more common. That is, they are the "normal" case we use all the time so I didn't want to have to say "Col". Rows mostly come up as the return type from the transpose operator `~v` but for the most part aren't seen explicitly.

`VectorView` and `Vector` both inherit from `VectorBase` and they have implicit conversions so a `VectorView` object can be passed to a `Vector` argument, for example. However, it would also be possible to have these methods return `Vector` but then we lose the implication that we're actually looking at the original data.